bigtop-packages/src/charm/zeppelin/layer-zeppelin/reactive/zeppelin.py

```diff
@@ -0,0 +1,158 @@
+
+    # Check status of Hive
+    if hive_joined and not hive_ready:
+        waiting_apps.append('hive')
+    elif hive_ready:
+        ready_apps.append('hive')
+
+    # Check status of Spark
+    if spark_joined and not spark_ready:
+        waiting_apps.append('spark')
+    elif spark_ready:
+        ready_apps.append('spark')
+
+    # Set appropriate status based on the apps we checked above
+    if waiting_apps:
+        hookenv.status_set('waiting',
+                           'waiting for: {}'.format(' & '.join(waiting_apps)))
+    elif ready_apps:
+        hookenv.status_set('active',
+                           'ready with: {}'.format(' & '.join(ready_apps)))
+    else:
+        hookenv.status_set('active', 'ready')
+
+
+@when('bigtop.available')
+@when_not('zeppelin.installed')
+def initial_setup():
+    hookenv.status_set('maintenance', 'installing zeppelin')
+    zeppelin = Zeppelin()
+    zeppelin.install()
+    zeppelin.setup_etc_env()
+    zeppelin.open_ports()
+    set_state('zeppelin.installed')
+    update_status()
+
+
+@when('zeppelin.installed')
+@when('hadoop.ready')
+@when_not('zeppelin.hadoop.configured')
+def configure_hadoop(hadoop):
+    zeppelin = Zeppelin()
+    zeppelin.configure_hadoop()
+    zeppelin.register_hadoop_notebooks()
+    set_state('zeppelin.hadoop.configured')
+
+
+@when('zeppelin.installed')
+@when_not('hadoop.ready')
+@when('zeppelin.hadoop.configured')
+def unconfigure_hadoop():
+    zeppelin = Zeppelin()
+    zeppelin.remove_hadoop_notebooks()
+    remove_state('zeppelin.hadoop.configured')
+
+
+@when('zeppelin.installed', 'hive.ready')
+def configure_hive(hive):
+    hive_ip = hive.get_private_ip()
+    hive_port = hive.get_port()
+    hive_url = 'jdbc:hive2://%s:%s' % (hive_ip, hive_port)
+    if data_changed('hive.connect', hive_url):
+        hookenv.status_set('maintenance', 'configuring hive')
+        zeppelin = Zeppelin()
+        zeppelin.configure_hive(hive_url)
+        set_state('zeppelin.hive.configured')
+        update_status()
+
+
+@when('zeppelin.installed', 'zeppelin.hive.configured')
+@when_not('hive.ready')
+def unconfigure_hive():
+    hookenv.status_set('maintenance', 'removing hive relation')
+    zeppelin = Zeppelin()
+    zeppelin.configure_hive('jdbc:hive2://:')
+    remove_state('zeppelin.hive.configured')
+    update_status()
+
+
+@when('zeppelin.installed', 'spark.ready')
+def configure_spark(spark):
+    master_url = spark.get_master_url()
+    if data_changed('spark.master', master_url):
+        hookenv.status_set('maintenance', 'configuring spark')
+        zeppelin = Zeppelin()
+        zeppelin.configure_spark(master_url)
+        set_state('zeppelin.spark.configured')
+        update_status()
+
+
+@when('zeppelin.installed', 'zeppelin.spark.configured')
+@when_not('spark.ready')
+def unconfigure_spark():
+    hookenv.status_set('maintenance', 'removing spark relation')
+    zeppelin = Zeppelin()
+    zeppelin.configure_spark(None)
```

Comment:
I think we should check for `hadoop.ready` here, and if that's true, use `yarn-client` as the spark master url instead of `None`.  I'm pretty sure zeppelin can run it's internal spark interpreter in yarn-client mode without any external spark master.

FWIW, bigtop defines `spark_master_url` as `yarn-client` by default:
https://github.com/apache/bigtop/blob/master/bigtop-deploy/puppet/modules/zeppelin/manifests/init.pp#L25